### ISSUE TITLE
rev-parse date reproducibility problems

### ISSUE THREAD

**Issue by daniel-brosche** [jacebrowning/gitman#188]
By using the rev-parse date approach to specify a dependency revision I figured out some inconsitencies over time.
Currently the rev-parse algorithm is based of the git rev-list command like:

```sh
git rev-list -n 1 --before="2018-11-29 13:35:58" master
```
By using this approach it can happen that the resolved sha-commit-id changes over time when afterwards a merged branch contains commits that are more close to the "before date".

For example:
Two weeks ago I specified a rev-parse-date  `master@{2018-11-29 13:35:58}` and at this point the resulted sha-commit-id was ae5cee39aa5867def113dabfe7ef9bd73f397f2b. The timestamp of this commit is 11/15/2018 07:35 AM. 
One week later I merged a feature branch into master and the same rev-parse-date resolves a different sha-commit-id because the rev-list command now also traverses all commits of the merged feature branch history. The resulted sha-commit-id was 0dc01018ef4b885857d7b920c70506a1cc74a74e and the timestamp of this commit is 11/15/2018 03:34 PM.

In my mind the resolved sha-commit-id should always be the same regardless of any merged branch in the future.
Otherwise the rev-parse-date cannot be safely used to reproducibly specify a revision of a dependency.

To achieve this, we need to adapt the internal used rev-list command by using `--first-parent` argument to only traverse the defined branch like:

```sh
git rev-list -n 1 --before="2018-11-29 13:35:58" --first-parent  master
``` 

Some more information can be found here:
https://stackoverflow.com/questions/41551560/git-what-is-the-meaning-of-first-parent-and-how-does-it-change-the-output-of
https://acassis.wordpress.com/2009/07/16/getting-git-revision-by-specific-date/

**Comment by jacebrowning** [comment]
Furthermore, I believe the timestamps can vary between machines.

I wouldn't recommend using `rev-parse` for any type of production deployment. It's there for cases when a project isn't versioned and simply need a vague time frame.

> Otherwise the rev-parse-date cannot be safely used to reproducibly specify a revision of a dependency.

The locking mechanism should still store a Git SHA.

**Comment by daniel-brosche** [comment]
I agree, I think it is helpful during development to import a certain state of the dependencies that are not tagged or even locked. Over this way some dependencies are by default fixed and other parts that are in the scope of the development initiative are imported as feature branches. This variable dependencies can also be locked (by buildserver) during development to mark the last verified stable state.
Therefore I see some benefits to combine this features for development purposes and not for production deployment.
Basically rev-parse does traverse the existing git history and I would like to adapt the traversal strategy as mentioned above. It should not result in a breaking change according to my tests.

**Comment by daniel-brosche** [comment]
Do you see any problems with this adaption?

**Comment by jacebrowning** [comment]
Nope. Feel free to submit a PR!
